Enhance iCal file exports

rtnpro opened the pull request:
This branch improves export of iCal files in the following ways:
- Allow export iCal file for a calendar event
- Add reminder logic in iCal file
- when downloading iCal info for a meeting, allow the option to add a reminder as well.

``NOTE``: This is work in progress and the PR has been created for review of ongoing work.

pypingou commented:
This is a nice start.
Had you kept #161 open I would have asked if you could adjust the unit-tests and I would have merged it, now I have to wait ;-)

rtnpro commented:
@pypingou I can reopen #161 then, and add unit tests in it :)

pypingou commented:
@rtnpro sure, works for me :)

rtnpro commented:
@pypingou Could you review it?

Changes:
- Add reminders in iCal file
- Allow option in meeting snippet to export meeting ical with/without reminders
- Allow configuring ical reminders from config file

rtnpro commented:
@pypingou if the above is OK, I will add tests for the same.

pypingou commented on fedocal.cfg.sample:
```diff
@@ -59,3 +59,6 @@ APP_URL = 'https://apps.fedoraproject.org/calendar/'
 ### problem. The email is sent to the address listed here.
 # EMAIL_ERROR = 'pingou@pingoured.fr'
+
+ICAL_REMINDERS = [{'days': -1}, {'hours': -1}, {'minutes': -5}]
```
We probably want to sync this with the default configuration and we should document a little more in the sample configuration

pypingou commented on fedocal/__init__.py:
```diff
@@ -621,7 +621,12 @@ def ical_calendar_meeting(meeting_id):
+        ical, meeting,
+        enable_reminder=flask.request.args.get('reminder') == '1',
+        reminders=APP.config.get('ICAL_REMINDERS') or [
+            {'days': -1}, {'hours': -1}, {'minutes': -5}
```
The ``or`` part can be dropped if we adjust ``default_config.py``

rtnpro replied:
I agree.

pypingou commented on fedocal/__init__.py:
```diff
@@ -621,7 +621,12 @@ def ical_calendar_meeting(meeting_id):
-    fedocallib.add_meeting_to_vcal(ical, meeting)
+    fedocallib.add_meeting_to_vcal(
+        ical, meeting,
+        enable_reminder=flask.request.args.get('reminder') == '1',
```
Not sure we should support only ``1`` here.

Might be an idea: https://github.com/fedora-infra/pkgdb2/blob/master/pkgdb2/api/packages.py#L1048-L1050

rtnpro replied:
``['1', 'true']`` makes sense. I will update the code to handle both the values.

pypingou commented on fedocal/fedocallib/__init__.py:
```diff
@@ -666,7 +666,7 @@ def retrieve_meeting_to_remind(session, offset=30):
 
 
-def add_meeting_to_vcal(ical, meeting):
+def add_meeting_to_vcal(ical, meeting, enable_reminder=False, reminders=[]):
```
Thinking out loud: do we need both arguments?
We could use only one: the reminder to set (either it is None, or there is one)

rtnpro replied:
Agreed :+1: So, we have a single kwarg: ``reminders==[]``.

pypingou replied:
``reminders=None`` would be better suited :)

pypingou commented on fedocal/fedocallib/__init__.py:
```diff
@@ -713,8 +717,19 @@ def add_meeting_to_vcal(ical, meeting):
 
+    if enable_reminder:
+        for trigger in (
+                timedelta(**reminder) for reminder in reminders):
```
So this means that by default we get 3 reminders for a single meeting?
Seems a little generous to me :-)

rtnpro replied:
I took inspiration for the default values from Google calendar. What default values would you prefer?

pypingou replied:
I'd rather have only 1 reminder for a meeting, I don't want to my phone/calendar app to keep buzzing me until I actually have to attend the meeting.

My default would be: remind me 10 or 15 minutes before the meeting

pypingou commented on fedocal/fedocallib/__init__.py:
```diff
@@ -722,6 +737,8 @@ def add_meetings_to_vcal(ical, meetings):
     :arg meetings: a list of fedocal.model.Meeting object to convert to
         iCal and add to the provided calendar.
+    :kwarg enable_reminder: a boolean telling whether to add reminder
+        for the meetings in the iCal object or not.
```
hm, what is ``enable_reminder`` used for in this method, did you forget a piece? :)

rtnpro replied:
Currently, it's not being used from the UI. But, it allows to add reminders for each meeting in a list of meetings. Oh! yes, I forgot the ``reminders`` param. Anyways, we are gonna replace ``enable_reminder`` with ``reminders`` param.

pypingou commented on fedocal/templates/default/view_meeting.html:
```diff
@@ -54,10 +54,16 @@ <h2 class="orange">{{ _('Meeting "%(name)s"', name=meeting.meeting_name) }}</h2>
+    <span>
+        <label for="ical-meeting-export-reminder-toggle">
+            <input id="ical-meeting-export-reminder-toggle" type="checkbox">
+            <em>{{ _('with reminder') }}</em>
```
Thinking on the idea mentioned above, should we have a drop-down list here instead of a checkbox?

rtnpro replied:
What will we have in the dropdown? Reminder before 1 day, 1 hour and 5 minutes? I think it will just complicate things in terms of UI. We can ask mizmo to help out with this decision :)

pypingou replied:
The drop down would give you the choice when you want to be reminded of the meeting

pypingou commented on fedocal/templates/default/view_meeting.html:
```diff
@@ -139,6 +145,17 @@ <h2 class="orange">{{ _('Meeting "%(name)s"', name=meeting.meeting_name) }}</h2>
+                    'href', $icalExportLink.attr('href') + '?reminder=1');
+            else
+                $icalExportLink.attr(
+                    'href', $icalExportLink.attr('href').split('?')[0]);
```
Neat :-)

rtnpro commented:
@pypingou Please review https://github.com/rtnpro/fedocal/commit/c99762f19aad0e592dec3fe7c1836c07eace4167. Please have a look at the commit summary for a summary of the changes made.

rtnpro commented:
@pypingou True! Currently the flow would allow setting reminders after the meeting if the user toys around the URL :) However, I wanted to keep the ``get_timestamp_from_str`` generic and wanted to have the policy: ``reminder can be set before the event`` in the view layer. However, values like: ``5 for 5 minutes, 60 for 1 hour and 60*24=1440 for 1 day`` seems simple and better. However, I am not able to think of a nice get parameter name for it. The param should explicitly mention the reminder is set at a time before the meeting. ``reminder_at`` needed ``-`` sign to indicate it. ``reminder_before`` won't need any sign before its value as the name itself says it all. But, I don't find the name ``reminder_before`` nice :\. What do you say?

pypingou commented on fedocal/__init__.py:
```diff
@@ -1030,7 +1038,9 @@ def view_meeting_page(meeting_id, full):
         editor=editor,
-        from_date=from_date)
+        from_date=from_date,
+        reminder_at_options=APP.config.get('ICAL_REMINDER_AT_OPTIONS') or []
```
No need for the ``or`` here, just finish the get:
``
reminder_at_options=APP.config.get('ICAL_REMINDER_AT_OPTIONS', [])
``
The second parameter is the default to return if the key is not in the dict ;-)

rtnpro replied:
@pypingou I know, but 
``reminder_at_options=APP.config.get('ICAL_REMINDER_AT_OPTIONS') or []``
handles the worst case scenario when
```
config[ICAL_REMINDER_AT_OPTIONS'] = None
```

It's unnecessary here. It's a side effect of writing fault tolerant code for E Commerce products to handle stupidity from some other sources ;). Here, it's unnecessary, I believe. I will fix it.

pypingou replied:
We shouldn't be able to have
``
config[ICAL_REMINDER_AT_OPTIONS'] = None
``
as it would mean someone edited the default_config.py instead of overriding it.

So +1 on not being tolerant to this, but in the worst case, I think I would prefer a ``if reminder_at_options`` lower down the stack.

rtnpro commented:
@pypingou I have added tests for this feature enhancement. Please review it.

pypingou commented on fedocal/templates/default/view_meeting.html:
```diff
@@ -54,10 +54,25 @@ <h2 class="orange">{{ _('Meeting "%(name)s"', name=meeting.meeting_name) }}</h2>
     <p>{{ _('Meeting organized by %(managers)s', managers=get_managers(meeting.meeting_manager)) }}</p>
 
-    <a href="{{ url_for('ical_calendar_meeting', meeting_id=meeting.meeting_id) }}"
+    <a id="ical-meeting-export" href="{{ url_for('ical_calendar_meeting', meeting_id=meeting.meeting_id) }}"
```
Might be good to place the id on the line below to keep (more or less) the 80 char rule :)

rtnpro replied:
Done!

pypingou commented on tests/test_flask.py:
```diff
@@ -432,6 +433,27 @@ def test_ical_meeting(self):
+                meeting_obj.meeting_time_stop).strftime(
+                    '%Y%m%dT%H%M%SZ')
+        )
+        output = self.app.get('/ical/calendar/meeting/2/?reminder_delta=60')
```
Could we test with a reminder_delta that is not an int? Also test with a reminder_delta that is not in the default choice (ie: 90)?

rtnpro replied:
OK :+1:

pypingou commented:
This is looking very nice and working nicely as well locally.

Two questions still and one remark: the tests are currently failing for me here :)
```
FAIL: Test the view_meeting_page function.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pingou/repos/gitrepo/fedocal/tests/test_flask.py", line 494, in test_view_meeting_page
    in output.data)
AssertionError: False is not true
```

rtnpro commented:
@pypingou I am looking into the above, and I will fix it ASAP.

rtnpro commented:
@pypingou I have moved the refactor related to static js files in another branch. I have also added the remainder tests that you'd mentioned. Please review it and let me know if it's ready for merge.

Please note that JS issues related to ical reminder UI elements persist for meeting details page. This has been already fixed in the branch for JS refactor.

pypingou commented:
This is looking really good!

I'm just going to wait for jenkins to kick in and then merge.

ralphbean commented:
Hooray!

:japanese_castle:  This was epic.  :european_castle:
